### PR TITLE
add `showDatesOutsideMonth` prop to `DatePicker`

### PR DESCRIPTION
**Pull request by mayank99**
## Changes

Added a new `showDatesOutsideMonth` prop. Addresses https://github.com/iTwin/iTwinUI/issues/1299#issuecomment-1917491982

The prop defaults to `true` to avoid a breaking change. In the next major version, we can flip it to `false`.

When `false`, this will render an empty div (to preserve layout). The CSS has also been adjusted to prevent mouse interaction (e.g. hover styling, cursor pointer).

There is one part I did not touch, i.e. showing an empty row in some months. It would look nicer without the empty row, but would cause a layout shift, which is arguably more unpleasant.
image

## Testing

Added unit test for the new prop.

I've updated all but one story to use `false`, since that's what we recommend. And I've only updated one html/css demo (mainly because it's so tedious but also because it doesn't matter) to remove content and attributes from dates outside month. This way, we are testing both permutations in both css and react, which I think is sufficient.

See preview: https://itwin.github.io/iTwinUI/1822/react/?story=datepicker--basic

image

Also tested that clicking on it the empty space does nothing, and keyboard navigation is unaffected.

## Docs

Added jsdocs and changesets.

I haven't updated the docsite yet, because it looks like the whole page needs to be re-written. I can do this in a separate PR, updating all demos.

### CODE REVIEW

File: packages/itwinui-react/src/core/DatePicker/DatePicker.tsx [hunk 1]
```diff
@@ -232,6 +232,14 @@ type DatePickerProps = {
...
    */
   applyBackground?: boolean;
+  /**
+   * Whether dates outside the current month should be displayed (in a muted text style).
+   *
+   * It is recommended to set this to false. Currently it defaults to true for backward compatibility.
+   *
+   * @default true
+   */
+  showDatesOutsideMonth?: boolean;
```
Comment (mayank99) [inline packages/itwinui-react/src/core/DatePicker/DatePicker.tsx:242]: Bikeshed comments welcome if you have better suggestions for naming this prop.

File: packages/itwinui-react/src/core/DatePicker/DatePicker.tsx [hunk 2]
```diff
@@ -665,6 +674,22 @@ export const DatePicker = React.forwardRef((props, forwardedRef) => {
...
+
+                  if (isOutsideMonth && !showDatesOutsideMonth) {
+                    return (
+                      <Box
+                        key={`day-${displayedMonthIndex}-${dayIndex}`}
+                        className={cx(
+                          getDayClass(weekDay),
+                          dayProps?.className,
+                        )}
+                        aria-hidden
```
[the remark this answers is not in the document]
Comment (mayank99) [reply packages/itwinui-react/src/core/DatePicker/DatePicker.tsx:688]: `inert` is typically used for preventing interactivity (preventing clicks, focus, etc). This is just a static div.

File: .changeset/silent-tools-exist.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@itwin/itwinui-css': minor
```
[the remark this answers is not in the document]
Comment (mayank99) [reply .changeset/silent-tools-exist.md:2]: I kept it as minor since `pointer-events` is kind of a big change. It's also going to be in a minor release anyway, because we have other minor changes pending release.

**Comment by AdamMeza-Bentley** [comment]
Given this new change, how would someone select dates that span outside of the month? is that not possible or am I missing the new method of selection?

File: apps/css-workshop/pages/date-picker.html [hunk 4]
```diff
@@ -113,60 +113,12 @@ <h2>Date</h2>
...
-              <div
-                class="iui-calendar-day-outside-month"
-                tabindex="0"
-                aria-label="day-31"
-                role="option"
-                aria-disabled="false"
-              >
-                31
-              </div>
+              <div class="iui-calendar-day-outside-month"></div>
```
[the remark this answers is not in the document]
Comment (mayank99) [reply apps/css-workshop/pages/date-picker.html:116]: Not necessary. These html pages are only meant to demonstrate styles.

Technically it would also not be necessary in React, because there is no content inside the div and no other a11y attributes. However, I've heard of cases where some screen-readers will announce "clickable" when event delegation is used, so `aria-hidden` prevents that.

**Review by mayank99 — COMMENTED** [review]
> Given this new change, how would someone select dates that span outside of the month? is that not possible or am I missing the new method of selection?

The month can still be changed using the `<`/`>` buttons at the top.

**Comment by AdamMeza-Bentley** [comment]
So, as a user, I would select the first date, then change the month and when I select a second date the datePicker would know im still in the middle of the selection process?

**Review by AdamMeza-Bentley — APPROVED** [review]
Looks good! I think the extra space, while maybe not ideal in the case of certain months, is preferable to size changes

File: packages/itwinui-react/src/core/DatePicker/DatePicker.tsx [hunk 5]
```diff
@@ -665,6 +674,22 @@ export const DatePicker = React.forwardRef((props, forwardedRef) => {
                 {weekDays.map((weekDay, dayIndex) => {
```
Comment (mayank99) [inline packages/itwinui-react/src/core/DatePicker/DatePicker.tsx:674]: One more consideration (non-blocking):

In some cases, using this prop results in empty space at the **top** rather than bottom. 

image

https://github.com/iTwin/iTwinUI/blob/e739baa6edb34c7b62a66f606b9e5aa781a51362/packages/itwinui-react/src/core/DatePicker/DatePicker.tsx#L342-L345

I'm not sure why it was designed this way originally, but I think it would make more sense to only show empty space at the bottom.

**Review by rohan-kadkol — APPROVED** [review]
